### PR TITLE
[WIP] [Prototype] Make portforward support udp

### PR DESCRIPTION
**Pull request by shiywang**
Fixes https://github.com/kubernetes/kubernetes/issues/47862
Proposal draft: https://github.com/kubernetes/community/pull/1575

still wip early stage, `kubectl` and `socat` side using udp connections, transmission still using spdy. need to consider how we split stream into udp packages.

- [ ] kubectl/client-go side support `--protocol=udp` flags, ListenUDPPacket
- [ ] kubelet side socat support udp, dockershim && rkt (use udp socket instead of stdin&&stdout since udp is not stream)
- [ ] make spdy stream transmit udp possible (all udp package share one stream connection, and use an customized udp protocol to distinguish)
- [ ] tests

@smarterclayton @ncdc @resouer any advice ? do I have to write a proposal ?

### CODE REVIEW

**Comment by ncdc** [comment]
I think at the very least this needs a new protocol version for port forward. A new client that requests UDP forwarding needs to be told by the server that it's not supported. We probably also need @kubernetes/api-reviewers and @kubernetes/api-approvers to weigh in.

I believe this may also need approval by/sponsorship from various SIGs (node, maybe network, maybe apps, maybe cli).

File: pkg/kubelet/dockershim/docker_streaming.go [hunk 1]
```diff
@@ -159,17 +170,7 @@ func attachContainer(client libdocker.Interface, containerID string, stdin io.Re
...
-	if err != nil {
-		return err
-	}
-
-	if !container.State.Running {
-		return fmt.Errorf("container not running (%s)", container.ID)
-	}
-
-	containerPid := container.State.Pid
+func protForwardTcp(containerPid int, port int32, stream io.ReadWriteCloser) error {
```
Comment (wjiangjay) [inline pkg/kubelet/dockershim/docker_streaming.go]: typo: prot

Comment (shiywang) [reply pkg/kubelet/dockershim/docker_streaming.go]: thanks!!

**Pull request closed without merging**